README.md
[![Build Status](https://travis-ci.org/pid/puid.png)](https://travis-ci.org/pid/puid)
[![NPM version](https://badge.fury.io/js/puid.png)](http://badge.fury.io/js/puid)

# puid - primary unique id

Generate an unique ID depending on time, machine and process for use in a distributed environment.


## puid - 24 characters uid

Each unique ID has 4 sections and has 24 chars, which are:

  i.e. he5fps6l2504cd1w3ag8ut8e // he5fps6l-2504cd-1w3a-g8ut8e
 
  - timestamp:    'he5fps6l'  // (8) Timestamp in microseconds - safe until 2059
  - machineId:    '2504cd'    // (6) First 6 chars from md5 of first external network interface or fallback to hostname
  - processId:    '1w3a'      // (4) pid
  - counter:      'g8ut8e'    // (6) High-resolution real time; nanoseconds

All values (except machineID) are converted to base36.

Why is the counter not really a counter? Because of collision, it's more likely that the same machine and process use (accidently!) two puid-objects (async) and generate an Id at the same microsecond with identical counter. That the same process will execute the counter function within the same nanosecond should be impossible (imo) - feedback is welcome.


## puid - short puid - 12-14 characters uid

Each unique ID has 2 sections and has 12 chars without nodeId or 14 chars with nodeId, which are:

    i.e. with 14-chars: aeby6ob5sso4zd // aeby6ob5sso4-zd

    - timestamp:    'aeby6ob5sso4'  // (12) Timestamp in seconds + hrtime counter
    - nodeId:       'zd'            // (2) nodeId

After the first puid release, I was testing for some shorter UIDs. In result, the short-puid was born ;-). Works only with node.js (depends on process.hrtime). The process.hrtime[1] gives you a counter (milliseconds+nanoseconds value from 1 to 1_000_000_000) not depending of the current timestamp; in result you can use short-puid (12 chars version) without any conflicts with multi-instances of puid in the same process or in different processes on the same host (tested with 20 parallel instances generating 4 billion puids). With multi nodes/hosts, you have to use the short-version with 14 chars which includes a nodeId.


### Configure short-puid

You have options :-) You can pass the epoch to start counting the timestamp, per default we start at "1999-06-07 03:00:00 pm GMT", and you can pass the nodeId. Take a look to the usage section for all possiblities.


## Usecase

Generate unique keys for e.g. Redis.


## Installation
[![NPM version](https://badge.fury.io/js/puid.png)](http://badge.fury.io/js/puid)

```bash
$ npm install puid
```


## Running tests

[![Build Status](https://travis-ci.org/pid/puid.png)](https://travis-ci.org/pid/puid)

```bash
$ npm test
```


## Usage

```js

var Puid = require('puid');
var puid;

// generate puid (long-version 24-chars) 

puid = new Puid();
console.log(puid.generate());   // hgqy29gr11cm2504cdf8rg7q


// generate puid (short-version 12-chars) without nodeId / **Shortcut**

puid = new Puid(true);
console.log(puid.generate());  // 3bqk9my968mc


// generate puid (short-version 14-chars) random nodeId / **Shortcut**

puid = new Puid(false);
console.log(puid.generate());  // 3bqk9my9buecut


// generate puid (short-version 14-chars) with nodeId string

puid = new Puid('JS');
console.log(puid.generate());   // 3bq1plt0vlycjs // 3bq1plt0vlyc->(js)


// generate puid (short-version 12-chars) with empty nodeId string

puid = new Puid('');
console.log(puid.generate());   // 3bq1plt1iiw4


// generate puid (short-version 14-chars) with nodeId in config object

puid = new Puid({
    nodeId: 'JS'
});
console.log(puid.generate());   // 3bq1plt1ljwgjs // 3bq1plt1ljwg->(js)


// generate puid (short-version 14-chars) with epoch in config object and random nodeId

puid = new Puid({
    epoch: '1980-01-01'
});
console.log(puid.generate());   // 801eeqggvq0w54


// generate puid (short-version 14-chars) with epoch and nodeId in config object

puid = new Puid({
    epoch: '1980-01-01',
    nodeId: 'JS'
});
console.log(puid.generate());   // 801eeqggyo00js // 801eeqggyo00->(js)


// generate puid (short-version 12-chars) with epoch and empty nodeId in config object

puid = new Puid({
    epoch: '2013-01-01',
    nodeId: ''
});
console.log(puid.generate());   // 036pqlnkkjk0


```


## [Changelog](https://raw.github.com/pid/puid/master/Changelog)

## License

[MIT](https://raw.github.com/pid/puid/master/LICENSE)

Copyright (c) 2013-2014 Sascha Droste <pid@posteo.net>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/pid/puid/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



Changelog
# Changelog

- v0.5.4 update dev dependencies
- v0.5.2 last version compatible with node v0.8.x
- v0.4.0 added short-puid, long-puid behavior is unchanged
- v0.5.0 changed datetime for default epoch value from "1999-07-02 03:00:00 pm" to "1999-06-07 03:00:00 pm GMT"; further infos see #7 (https://github.com/pid/puid/issues/7)
    

test/test.js
var Puid = require('../lib/puid'),
  pid = new Puid(),
  should = require('should');

var oldpid = pid.generate();

describe('lib/puid.js', function() {

  describe('getTimestamp', function() {
    it('should be a string with 8 chars', function() {
      pid.getTimestamp().should.be.a('string');
      pid.getTimestamp().should.have.length(8);
    });
    it('should be a sexhex value', function() {
      pid.getTimestamp().should.match(/[a-z0-9]/);
    });
    it('two request should not be equal', function() {
      pid.getTimestamp().should.not.eql(oldpid);
    });
  });

  describe('getMachineId', function() {
    it('should be a string with 6 chars', function() {
      pid.getMachineId().should.be.a('string');
      pid.getMachineId().should.have.length(6);
    });
    it('should be 089776', function() {
      pid.getMachineId(m0).should.be.eql('089776');
    });
    it('should be a093d8', function() {
      pid.getMachineId(m1).should.be.eql('a093d8');
    });
    it('should be dcfbe5', function() {
      pid.getMachineId(m2).should.be.eql('dcfbe5');
    });
    it('should be a sexhex value', function() {
      pid.getMachineId().should.match(/[a-z0-9]/);
    });
    it('should use consistent fallback value', function() {
      pid.getMachineId(m3, 'localhost').should.be.equal('d13e79');
      pid.getMachineId(m3, 'localhost').should.have.length(6);
    });
    it('should use consistent fallback value without external network interfaces', function() {
      pid.getMachineId({}, 'localhost').should.be.equal('d13e79');
      pid.getMachineId({}, 'localhost').should.have.length(6);
    });
  });

  describe('getProcessId', function() {
    it('should be a string with 4 chars', function() {
      pid.getProcessId().should.be.a('string');
      pid.getProcessId().should.have.length(4);
    });
    it('should be a sexhex value', function() {
      pid.getProcessId().should.match(/[a-z0-9]/);
    });
    it('two request should be equal', function() {
      pid.getProcessId().should.equal(pid.getProcessId());
    });
    it('internal function process.pid is available', function() {
      (process).should.be.a('object');
      (typeof process.pid).should.not.be.equal('undefined');
    });
  });

  describe('getCounter', function() {
    it('should be a string with 6 chars', function() {
      pid.getCounter().should.be.a('string');
      pid.getCounter().should.have.length(6);
    });
    it('should be a sexhex value', function() {
      pid.getCounter().should.match(/[a-z0-9]/);
    });
    it('two request should not be equal', function() {
      pid.getCounter().should.not.equal(pid.getCounter());
    });
    it('internal function process.hrtime is available', function() {
      (process).should.be.a('object');
      (typeof process.hrtime).should.not.be.equal('undefined');
    });
  });

  describe('generate pId', function() {
    it('should be a string with 24 chars', function() {
      pid.generate().should.be.a('string');
      pid.generate().should.have.length(24);
    });
    it('should be a sexhex value', function() {
      pid.generate().should.match(/[a-z0-9]/);
    });
    it('two request should not be equal', function() {
      pid.generate().should.not.equal(pid.generate());
    });
  });

  describe('generate base36 string', function() {
    it('should be result in sexhex values', function() {
      pid.toBase36String('22').should.be.equal('0m');
      pid.toBase36String('22', 2).should.be.equal('0m');
      pid.toBase36String('22', 3).should.be.equal('00m');
      pid.toBase36String('22', 'foobar').should.be.equal('0m');
      pid.toBase36String('1000000', 5).should.be.equal('0lfls');
      pid.toBase36String('10000000000', 8).should.be.equal('04ldqpds');
      pid.toBase36String('10000000000', 9).should.be.equal('004ldqpds');
      pid.toBase36String('1000000', 4).should.be.equal('lfls');
    });
    it('should throw an error', function() {
      (function() {
        pid.toBase36String('foobar', 4);
      }).should.throwError(/foobar/);
    });
    it('should not throw an error, use default padding', function() {
      (function() {
        pid.toBase36String('1000', 'foobar');
      }).should.not.throwError();
    });
  });
  describe('check package.json', function() {
    it('if format is valid', function() {
      require('../package.json');
    });
  });
});


var m0 = {
  lo0: [{
    address: 'fe80::1',
    family: 'IPv6',
    internal: true
  }, {
    address: '127.0.0.1',
    family: 'IPv4',
    internal: true
  }, {
    address: '::1',
    family: 'IPv6',
    internal: true
  }],
  en1: [{
    address: 'fe80::7bc3:8ac3:ea45:c34a',
    family: 'IPv6',
    internal: false
  }, {
    address: '192.168.3.22',
    family: 'IPv4',
    internal: false
  }],
  vnic0: [{
    address: '10.211.55.2',
    family: 'IPv4',
    internal: false
  }],
  vnic1: [{
    address: '10.37.129.2',
    family: 'IPv4',
    internal: false
  }]
}

var m1 = {
  Drahtlosnetzwerkverbindung:   [{
    address: 'fe80::f322:3e0f:ced6:5b83',
    family: 'IPv6',
    internal: false
  }, {
    address: '192.168.2.103',
    family: 'IPv4',
    internal: true
  }],
   
  'LAN-Verbindung':   [{
    address: 'fe80::93d3:34c9:7e0a:edc3',
    family: 'IPv6',
    internal: false
  }, {
    address: '192.168.2.10',
    family: 'IPv4',
    internal: true
  }],
   
  'VMware Network Adapter VMnet1':   [{
    address: 'fe80::1c3e:c641:340e:ea23',
    family: 'IPv6',
    internal: false
  }, {
    address: '169.254.6.66',
    family: 'IPv4',
    internal: true
  }],
   
  'VMware Network Adapter VMnet8':   [{
    address: 'fe80::6465:562d:2b25:34ef',
    family: 'IPv6',
    internal: false
  }, {
    address: '192.168.142.1',
    family: 'IPv4',
    internal: true
  }],
   
  'Loopback Pseudo-Interface 1':   [{
    address: '::1',
    family: 'IPv6',
    internal: true
  }, {
    address: '127.0.0.1',
    family: 'IPv4',
    internal: true
  }]
};

var m2 = {
  Ethernet:   [{
    address: 'fe80::cabf:e927:f866:8f3e',
    family: 'IPv6',
    internal: false
  }, {
    address: '10.16.16.22',
    family: 'IPv4',
    internal: false
  }],
  'Loopback Pseudo-Interface 1':   [{
    address: '::1',
    family: 'IPv6',
    internal: true
  }, {
    address: '127.0.0.1',
    family: 'IPv4',
    internal: true
  }],
  'LAN-Verbindung* 11':   [{
    address: '2001:3450:5af1:29fe:375a:fb8f:e54f:af39',
    family: 'IPv6',
    internal: false
  }, {
    address: 'fe80::8493:ab4f:e54f:cf39',
    family: 'IPv6',
    internal: false
  }]
};

var m3 = {
  Ethernet:   [{
    address: 'fe80::cabf:e927:f866:8f3e',
    family: 'IPv6',
    internal: true
  }, {
    address: '10.16.16.22',
    family: 'IPv4',
    internal: true
  }],
  'Loopback Pseudo-Interface 1':   [{
    address: '::1',
    family: 'IPv6',
    internal: true
  }, {
    address: '127.0.0.1',
    family: 'IPv4',
    internal: true
  }],
  'LAN-Verbindung* 11':   [{
    address: '2001:3450:5af1:29fe:375a:fb8f:e54f:af39',
    family: 'IPv6',
    internal: true
  }, {
    address: 'fe80::8493:ab4f:e54f:cf39',
    family: 'IPv6',
    internal: true
  }]
};

test/test-long-puid.js
/* jshint -W015: true, -W030 */
var Puid = require('../lib/puid');
var pid = new Puid();

var oldpid = pid.generate();

var m0 = {
    lo0: [{
        address: 'fe80::1',
        family: 'IPv6',
        internal: true
    }, {
        address: '127.0.0.1',
        family: 'IPv4',
        internal: true
    }, {
        address: '::1',
        family: 'IPv6',
        internal: true
    }],
    en1: [{
        address: 'fe80::7bc3:8ac3:ea45:c34a',
        family: 'IPv6',
        internal: false
    }, {
        address: '192.168.3.22',
        family: 'IPv4',
        internal: false
    }],
    vnic0: [{
        address: '10.211.55.2',
        family: 'IPv4',
        internal: false
    }],
    vnic1: [{
        address: '10.37.129.2',
        family: 'IPv4',
        internal: false
    }]
};

var m1 = {
    'Drahtlosnetzwerkverbindung': [{
        address: 'fe80::f322:3e0f:ced6:5b83',
        family: 'IPv6',
        internal: false
    }, {
        address: '192.168.2.103',
        family: 'IPv4',
        internal: true
    }],
    'LAN-Verbindung':  [{
        address: 'fe80::93d3:34c9:7e0a:edc3',
        family: 'IPv6',
        internal: false
    }, {
        address: '192.168.2.10',
        family: 'IPv4',
        internal: true
    }],

    'VMware Network Adapter VMnet1': [{
        address: 'fe80::1c3e:c641:340e:ea23',
        family: 'IPv6',
        internal: false
    }, {
        address: '169.254.6.66',
        family: 'IPv4',
        internal: true
    }],

    'VMware Network Adapter VMnet8':  [{
        address: 'fe80::6465:562d:2b25:34ef',
        family: 'IPv6',
        internal: false
    }, {
        address: '192.168.142.1',
        family: 'IPv4',
        internal: true
    }],
    'Loopback Pseudo-Interface 1':  [{
        address: '::1',
        family: 'IPv6',
        internal: true
    }, {
        address: '127.0.0.1',
        family: 'IPv4',
        internal: true
    }]
};

var m2 = {
    Ethernet:  [{
        address: 'fe80::cabf:e927:f866:8f3e',
        family: 'IPv6',
        internal: false
    }, {
        address: '10.16.16.22',
        family: 'IPv4',
        internal: false
    }],
    'Loopback Pseudo-Interface 1':  [{
        address: '::1',
        family: 'IPv6',
        internal: true
    }, {
        address: '127.0.0.1',
        family: 'IPv4',
        internal: true
    }],
    'LAN-Verbindung* 11':  [{
        address: '2001:3450:5af1:29fe:375a:fb8f:e54f:af39',
        family: 'IPv6',
        internal: false
    }, {
        address: 'fe80::8493:ab4f:e54f:cf39',
        family: 'IPv6',
        internal: false
    }]
};

var m3 = {
    Ethernet:  [{
        address: 'fe80::cabf:e927:f866:8f3e',
        family: 'IPv6',
        internal: true
    }, {
        address: '10.16.16.22',
        family: 'IPv4',
        internal: true
    }],
    'Loopback Pseudo-Interface 1':  [{
        address: '::1',
        family: 'IPv6',
        internal: true
    }, {
        address: '127.0.0.1',
        family: 'IPv4',
        internal: true
    }],
    'LAN-Verbindung* 11':  [{
        address: '2001:3450:5af1:29fe:375a:fb8f:e54f:af39',
        family: 'IPv6',
        internal: true
    }, {
        address: 'fe80::8493:ab4f:e54f:cf39',
        family: 'IPv6',
        internal: true
    }]
};

describe('test long puid', function() {

    describe('getTimestamp', function() {

        it('should be a string with 8 chars', function() {
            pid.getTimestamp()
                .should.be.a.String;
            pid.getTimestamp()
                .should.have.length(8);
        });

        it('should be a sexhex value', function() {
            pid.getTimestamp()
                .should.match(/[a-z0-9]/);
        });

        it('two request should not be equal', function() {
            pid.getTimestamp()
                .should.not.eql(oldpid);
        });
    });

    describe('getNodeId', function() {

        it('should be a string with 6 chars', function() {
            pid.getNodeId()
                .should.be.a.String;
            pid.getNodeId()
                .should.have.length(6);
        });

        it('should be 089776', function() {
            pid.getNodeId(m0)
                .should.be.eql('089776');
        });

        it('should be a093d8', function() {
            pid.getNodeId(m1)
                .should.be.eql('a093d8');
        });

        it('should be dcfbe5', function() {
            pid.getNodeId(m2)
                .should.be.eql('dcfbe5');
        });

        it('should be a sexhex value', function() {
            pid.getNodeId()
                .should.match(/[a-z0-9]/);
        });

        it('should use consistent fallback value', function() {
            pid.getNodeId(m3, 'localhost')
                .should.be.equal('d13e79');
            pid.getNodeId(m3, 'localhost')
                .should.have.length(6);
        });

        it('should use consistent fallback value without external network interfaces', function() {
            pid.getNodeId({}, 'localhost')
                .should.be.equal('d13e79');
            pid.getNodeId({}, 'localhost')
                .should.have.length(6);
        });
    });

    describe('getProcessId', function() {

        it('should be a string with 4 chars', function() {
            pid.getProcessId()
                .should.be.a.String;
            pid.getProcessId()
                .should.have.length(4);
        });

        it('should be a sexhex value', function() {
            pid.getProcessId()
                .should.match(/[a-z0-9]/);
        });

        it('two request should be equal', function() {
            pid.getProcessId()
                .should.equal(pid.getProcessId());
        });

        it('internal function process.pid is available', function() {
            (process).should.be.an.Object;
            (typeof process.pid)
                .should.not.be.equal('undefined');
        });
    });

    describe('getCounter', function() {

        it('should be a string with 6 chars', function() {
            pid.getCounter()
                .should.be.a.String;
            pid.getCounter()
                .should.have.length(6);
        });

        it('should be a sexhex value', function() {
            pid.getCounter()
                .should.match(/[a-z0-9]/);
        });

        it('two request should not be equal', function() {
            pid.getCounter()
                .should.not.equal(pid.getCounter());
        });

        it('internal function process.hrtime is available', function() {
            (process).should.be.an.Object;
            (typeof process.hrtime)
                .should.not.be.equal('undefined');
        });
    });

    describe('generate pId', function() {

        it('should be a string with 24 chars', function() {
            pid.generate()
                .should.be.a.String;
            pid.generate()
                .should.have.length(24);
        });

        it('should be a sexhex value', function() {
            pid.generate()
                .should.match(/[a-z0-9]/);
        });

        it('two request should not be equal', function() {
            pid.generate()
                .should.not.equal(pid.generate());
        });
    });

    describe('generate base36 string', function() {

        it('should be result in sexhex values', function() {
            pid.toBase36String('22')
                .should.be.equal('0m');
            pid.toBase36String('22', 2)
                .should.be.equal('0m');
            pid.toBase36String('22', 3)
                .should.be.equal('00m');
            pid.toBase36String('22', 'foobar')
                .should.be.equal('0m');
            pid.toBase36String('1000000', 5)
                .should.be.equal('0lfls');
            pid.toBase36String('10000000000', 8)
                .should.be.equal('04ldqpds');
            pid.toBase36String('10000000000', 9)
                .should.be.equal('004ldqpds');
            pid.toBase36String('1000000', 4)
                .should.be.equal('lfls');
        });

        it('should throw an error', function() {
            ! function() {
                pid.toBase36String('foobar', 4);
            }.should.throwError(/foobar/);
        });

        it('should not throw an error, use default padding', function() {
            ! function() {
                pid.toBase36String('1000', 'foobar');
            }.should.not.throwError();
        });
    });

    describe('check package.json', function() {
        it('if format is valid', function() {
            require('../package.json');
        });
    });
});


test/test-short-puid.js
/* jshint -W030 */

var Puid = require('../lib/puid');

function getEpoch(years) {
    var now = new Date()
        .valueOf();

    now = now - 86400000 * 365 * years;

    var epoch = new Date(now);

    return epoch.getFullYear() + "-" + (epoch.getMonth() + 1) + "-" + epoch.getDate() + " " + epoch.getHours() + ":" + epoch.getMinutes() + ":" + epoch.getSeconds();
}

describe('test short puid', function() {

    it('with explicit nodeId as string', function() {

        var pid = new Puid('ab');

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(14);
        pid.generate()
            .should.match(/ab$/);
        pid.generate()
            .should.match(/[a-z0-9]+/);
    });

    it('with explicit nodeId as uppercase string', function() {

        var pid = new Puid('AB');

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(14);
        pid.generate()
            .should.match(/ab$/);
        pid.generate()
            .should.match(/[a-z0-9]+/);
    });

    it('with random nodeId', function() {

        var pid = new Puid('random');

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(14);
        pid.generate()
            .should.match(/[a-z0-9]+/);

    });

    it('with empty nodeId string', function() {

        var pid = new Puid('');

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(12);
        pid.generate()
            .should.match(/[a-z0-9]+/);

    });

    it('with config object set empty nodeId', function() {

        var pid = new Puid({
            nodeId: ''
        });

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(12);
        pid.generate()
            .should.match(/[a-z0-9]+/);

    });

    it('with config object set nodeId', function() {

        var pid = new Puid({
            nodeId: "tt"
        });

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(14);
        pid.generate()
            .should.match(/tt$/);
        pid.generate()
            .should.match(/[a-z0-9]+/);
    });

    it('with config object set epoch to 1 year ago', function() {

        var pid = new Puid({
            epoch: getEpoch(1)
        });

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(14);
        pid.generate()
            .should.match(/^08miku/);
        pid.generate()
            .should.match(/[a-z0-9]+/);

    });

    it('with config object set epoch to 11 years ago', function() {

        var pid = new Puid({
            epoch: getEpoch(11)
        });

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(14);
        pid.generate()
            .should.match(/^2mvod/);
        pid.generate()
            .should.match(/[a-z0-9]+/);

    });

    it('with config object set empty nodeId and epoch to 11 year ago', function() {

        var pid = new Puid({
            nodeId: '',
            epoch: getEpoch(11)
        });

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(12);
        pid.generate()
            .should.match(/^2mvod/);
        pid.generate()
            .should.match(/[a-z0-9]+/);

    });

    it('with param "true"', function() {

        var pid = new Puid(true);

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(12);
        pid.generate()
            .should.match(/[a-z0-9]+/);

    });

    it('with param "false"', function() {

        var pid = new Puid(false);

        pid.generate()
            .should.be.a.String;
        pid.generate()
            .should.have.length(14);
        pid.generate()
            .should.match(/[a-z0-9]+/);

    });

    it('should have specific default epoch value"', function() {

        var pid = new Puid(false);

        pid.config.epoch.should.equal("1999-06-07 03:00:00 pm GMT");
    });
});

//


Gruntfile.js
module.exports = function(grunt) {
    'use strict';

    // Project configuration.
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),

        jshint: {
            options: {
                jshintrc: '.jshintrc'
            },

            src: ['Gruntfile.js', '<%= sourceFiles %>', '<%= testFiles %>']
        },

        watch: {
            files: ['Gruntfile.js', 'lib/**/*.js', 'test/**/*.js', 'package.json'],
            tasks: ['jshint', 'mocha']
        },

        bumpup: ['package.json'],

        release: {
            options: {
                bump: false, //default: true
                tagName: 'v<%= version %>', //default: '<%= version %>'
                commitMessage: 'release v<%= version %>', //default: 'release <%= version %>'
                tagMessage: 'tagging version v<%= version %>' //default: 'Version <%= version %>'
            }
        },

        // files
        buildSourceFile: 'lib/index.js',
        sourceFiles: 'lib/**/*.js',
        testFiles: 'test/**/*.js'
    });

    grunt.registerTask('mocha', 'run mocha', function() {
        var done = this.async();
        require('child_process').exec('mocha', function(err, stdout) {
            grunt.log.write(stdout);
            done(err);
        });
    });

    grunt.event.on('watch', function(action, filepath) {
        grunt.log.writeln(filepath + ' has ' + action);
    });

    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-release');
    grunt.loadNpmTasks('grunt-bumpup');

    // Default task(s).
    grunt.registerTask('default', ['jshint', 'mocha']);

};


examples/generatePid.js
var Puid = require('../lib/puid');

var puid;

// ####

console.log("\ngenerate puid (long-version)\n");

puid = new Puid();

for (var i = 0; i < 3; i++) {
    console.log(puid.generate());
}

// ####

console.log("\ngenerate puid (short-version) with nodeId string\n");

puid = new Puid('JS');

for (var i = 0; i < 3; i++) {
    console.log(puid.generate());
}

// ####

console.log("\ngenerate puid (short-version) with empty nodeId string\n");

puid = new Puid('');

for (var i = 0; i < 3; i++) {
    console.log(puid.generate());
}

// ####

console.log("\ngenerate puid (short-version) with nodeId in config object \n");

puid = new Puid({
    nodeId: 'JS'
});

for (var i = 0; i < 3; i++) {
    console.log(puid.generate());
}

// ####

console.log("\ngenerate puid (short-version) with epoch in config object and random nodeId\n");

puid = new Puid({
    epoch: '1980-01-01'
});

for (var i = 0; i < 3; i++) {
    console.log(puid.generate());
}

// ####

console.log("\ngenerate puid (short-version) with epoch and nodeId in config object\n");

puid = new Puid({
    epoch: '1980-01-01',
    nodeId: 'JS'
});

for (var i = 0; i < 3; i++) {
    console.log(puid.generate());
}

// ####

console.log("\ngenerate puid (short-version) with epoch and empty nodeId in config object\n");

puid = new Puid({
    epoch: '2013-01-01',
    nodeId: ''
});

for (var i = 0; i < 3; i++) {
    console.log(puid.generate());
}

// ####

console.log("\ngenerate puid (short-version 12-chars) without nodeId / Shortcut new Puid(true)!\n");

puid = new Puid(true);

for (var i = 0; i < 3; i++) {
    console.log(puid.generate());
}

// ####

console.log("\ngenerate puid (short-version 14-chars) random nodeId / Shortcut new Puid(false)!\n");

puid = new Puid(false);

for (var i = 0; i < 3; i++) {
    console.log(puid.generate());
}

//

lib/puid.js
var os = require('os');
var crypto = require('crypto');

var Puid = function Piud(options) {

    "use strict";

    this.config = {
        "nlen": options === void 0 ? 6 : 2, // machine identifier for long or short puid,
        "plen": 4, // process id
        "tlen": options === void 0 ? 8 : 12, // timestamp length for long or short puid
        "clen": 6, // counter/hrtime for long puid
        "epoch": "1999-06-07 03:00:00 pm GMT" // epoch start for short puid
    };

    if (options === void 0) {

        // long puid
        this.type = "long";

        this.nodeId = this.getNodeId();
        this.processId = this.getProcessId();

    } else {

        // short puid
        this.type = "short";

        if (typeof options === 'string' && options.length <= 2) {
            this.nodeId = options;
        } else {

            if (typeof options === 'object') {

                if (options.nodeId !== void 0 && typeof options.nodeId === 'string' && options.nodeId.length <= 2) {
                    this.nodeId = options.nodeId;
                }

                if (options.epoch !== void 0 && typeof options.epoch === 'string' && options.epoch.length >= 4) {
                    if (isNaN(Date.parse(options.epoch))) {
                        throw new Error("epoch date format error");
                    }
                    this.config.epoch = options.epoch;
                }
            }
        }

        // random nodeId
        if (options !== true && this.nodeId === void 0) {
            this.nodeId = this.toBase36String(parseInt(Math.random() * (Math.pow(36, 2) - 1), 10), this.config.nlen);
        } else {
            if (options === true) {
                this.nodeId = '';
            }
        }

        this.nodeId = this.nodeId && this.nodeId.toLowerCase()
            .trim();
    }
};

Puid.prototype.generate = function generate() {

    if (this.type[0] === "l") {
        return this.getTimestamp() + this.processId + this.nodeId + this.getCounter();
    } else {
        return this.getNanos() + this.nodeId;
    }

};

Puid.prototype.getTimestamp = function getTimestamp() {

    return this.toBase36String(Date.now(), this.config.tlen);
};

Puid.prototype.getNanos = function getNanos() {

    var hrt = '' + process.hrtime()[1];

    while (hrt.length < 9) {
        hrt = "0" + hrt;
    }

    var time = parseInt(Date.now() / 1000, 10) - parseInt(Date.parse(this.config.epoch) / 1000, 10);

    return this.toBase36String(time + '' + hrt, this.config.tlen);
};

Puid.prototype.getNodeId = function getNodeId(networkInterfaces, fallback) {

    var i, v, value;

    networkInterfaces = networkInterfaces || os.networkInterfaces() || {};

    if (!fallback) {
        /* jshint -W015: true */
        interfaces: for (i in networkInterfaces) {
            for (v in i) {
                if (networkInterfaces[i][v] && networkInterfaces[i][v].address.length && !networkInterfaces[i][v].internal) {
                    value = networkInterfaces[i][v].address;
                    break interfaces;
                }
            }
        }
    }

    return crypto.createHash('md5')
        .update(fallback || value || os.hostname(), 'utf8')
        .digest('hex')
        .slice(-this.config.nlen);
};

Puid.prototype.getProcessId = function getProcessId() {

    return this.toBase36String(process.pid, this.config.plen);
};

Puid.prototype.getCounter = function getCounter() {

    return this.toBase36String(process.hrtime()[1], this.config.clen);
};

Puid.prototype.toBase36String = function toBase36String(value, padding) {

    if (isNaN(value)) {
        throw new Error('error in toBase36String: value is NaN: \"' + value + '\"');
    }
    padding = padding && !isNaN(padding) ? padding : 2;

    var hex = Number(Math.abs(value))
        .toString(36);

    while (hex.length < padding) {
        hex = "0" + hex;
    }

    return hex.substr(0, padding);
};

module.exports = exports = Puid;

//
